packages/app-folders/src/components/Tree/List/index.tsx

```diff
@@ -0,0 +1,214 @@
+
+import { useFolders } from "~/hooks";
+
+import { DndItemData, FolderItem } from "~/types";
+
+const ROOT_ID = "root";
+
+const createTreeData = (
+    folders: FolderItem[] = [],
+    focusedNodeId?: string
+): NodeModel<DndItemData>[] => {
+    return folders.map(item => {
+        const { id, parentId, name, slug, type, createdOn, createdBy } = item;
+
+        return {
+            id,
+            parent: parentId || ROOT_ID,
+            text: name,
+            droppable: true,
+            data: {
+                id,
+                name,
+                slug,
+                parentId,
+                type,
+                createdOn,
+                createdBy,
+                isFocused: focusedNodeId === id
+            }
+        };
+    });
+};
+
+const createInitialOpenList = (
+    folders: FolderItem[] = [],
+    openIds: NodeModel<DndItemData>["id"][] = [],
+    focusedId?: string
+): InitialOpen | undefined => {
+    const focusedFolder = folders.find(folder => {
+        return folder.id === focusedId;
+    });
+
+    if (!focusedId || !focusedFolder?.parentId) {
+        return openIds;
+    }
+
+    const findParents = (
+        acc: string[],
+        folder: FolderItem,
+        index: number,
+        folders: FolderItem[]
+    ): string[] => {
+        if (folder.parentId && acc.some(el => el === folder.id)) {
+            acc.push(folder.parentId);
+
+            const newArr = folders.filter((_, i) => {
+                return i !== index;
+            });
+
+            return newArr.reduce(findParents, acc);
+        }
+
+        return acc;
+    };
+
+    const result = folders.reduce(findParents, [focusedId]);
+
+    return [...new Set([...result, ...openIds])];
+};
+
+type Props = {
+    type: string;
+    folders: FolderItem[];
+    focusedFolderId?: string;
+    onFolderClick: (data: NodeModel<DndItemData>["data"]) => void;
+    onDragStart: () => void;
+    onDragEnd: () => void;
+};
+
+export const List = ({
+    type,
+    folders,
+    onFolderClick,
+    focusedFolderId,
+    onDragStart,
+    onDragEnd
+}: Props) => {
+    const { updateFolder } = useFolders(type);
+    const { showSnackbar } = useSnackbar();
+    const [treeData, setTreeData] = useState<NodeModel<DndItemData>[]>([]);
+    const [initialOpenList, setInitialOpenList] = useState<undefined | InitialOpen>(undefined);
+    const [openFolderIds, setOpenFolderIds] = useState<NodeModel<DndItemData>["id"][]>([]);
+    const [updateDialogOpen, setUpdateDialogOpen] = useState<boolean>(false);
+    const [deleteDialogOpen, setDeleteDialogOpen] = useState<boolean>(false);
+    const [selectedFolder, setSelectedFolder] = useState<FolderItem>();
+
+    useDeepCompareEffect(() => {
+        if (folders) {
+            setTreeData(createTreeData(folders, focusedFolderId));
+            setInitialOpenList(createInitialOpenList(folders, openFolderIds, focusedFolderId));
+        }
+    }, [Object.assign({}, folders), focusedFolderId]);
```

Comment:
Hm, can you explain `Object.assign({}, folders)` here?

Also, if really needed, `[{ ...folders }, focusedFolderId]` might be a bit easier to read.

Reply:
Yeah, that's enough 😄